### DISCUSSION TITLE
Coveo Usage Analytics in Angular

### DISCUSSION THREAD

**Discussion by akaustav** [coveo/cli#989] [Q&A] [answered]
Hi,
I am not sure if the following question(s) belong in here or in a separate repository - so trying out here first.

After moving into Coveo Headless (not Atomic) in our Angular 11 application, the search term suggestions / auto-completions stopped working. After doing some research, it seems we might be missing [Usage Analytics](https://docs.coveo.com/en/headless/latest/usage/headless-usage-analytics/) in our implementation with Coveo Headless. I have a few questions around Usage Analytics.

1. Does `@coveo/cli` have any way to generate an example Coveo Headless based Angular application with Usage Analytics? If not, is there a sample Angular (preferably > Angular v10) application that can be used as a reference?

1. [This page](https://docs.coveo.com/en/headless/latest/usage/headless-usage-analytics/headless-view-events/#send-view-events-with-react) provides some information around handling coveo analytics with React applications.

1. Judging from PR [#523](https://github.com/coveo/cli/pull/523), is Usage Analytics being deprecated in favor of [@amplitude/node](https://www.npmjs.com/package/@amplitude/node)?

Ameet

**Comment by y-lakhdar** [comment]
Hi Ameet,

You can run the `ui:create:angular` command which will bootstrap a Coveo search app using Angular. 
This demo project will include a searchbox as well as a query suggestions leveraging Usage Analytics data.

If you are still not getting suggestions, please check that you are using the right Search Hub from which your [Query Suggest Model](https://docs.coveo.com/en/3386/leverage-machine-learning/about-query-suggestions-qs) learns from.
![image](https://user-images.githubusercontent.com/12199712/196467845-31d323f1-f71b-41a2-882c-792b7eb20aad.png)

> Judging from PR https://github.com/coveo/cli/pull/523, is Usage Analytics being deprecated in favor of [@amplitude/node](https://www.npmjs.com/package/@amplitude/node)?

You don't need to worry about this since it has nothing to do with your implementation. We used Coveo analytics internally to track CLI usage but decided to change to Amplitude.

**Reply by akaustav** [reply]
Thank you for your response, @y-lakhdar. My responses are below:

> You can run the `ui:create:angular` command which will bootstrap a Coveo search app using Angular. This demo project will include a searchbox as well as a query suggestions leveraging Usage Analytics data.

I was able to run the demo project in my local connected to the correct `searchHub` in our coveo account. No query suggestions show up.

> If you are still not getting suggestions, please check that you are using the right Search Hub from which your [Query Suggest Model](https://docs.coveo.com/en/3386/leverage-machine-learning/about-query-suggestions-qs) learns from.

We are definitely using the correct Search Hub that the Query Suggest Model is set to learn from. (I'll edit this post and attach some screenshots shortly)

> You don't need to worry about this since it has nothing to do with your implementation. We used Coveo analytics internally to track CLI usage but decided to change to Amplitude.

Understood. Thanks for clarifying this confusion.
I'll follow up with another response with things that I have learnt so far.

**Reply by akaustav** [reply]
@y-lakhdar - Apologies - I had a busy day yesterday - forgot to post screenshots that I had promised in my previous comment. I am attaching three screenshots here:

## Screenshot 1

Here is a screenshot of the `CoveoEngineService` service in the Angular UI layer employing `@coveo/headless` - that uses `searchHub: 'website-full-search'`.

![image](https://user-images.githubusercontent.com/11030959/197876762-6d773e73-877d-4b59-a7f7-ac82c68dfe1a.png)

## Screenshot 2

Here is a screenshot of the API middleware generating the token for searches - that uses `searchHub: 'website-full-search'`. As you can see, we do NOT use `@coveo/platform-client.PlatformClient`. Instead, we fetch the token via a REST over HTTP call (via a library called `axios`) to one of the coveo endpoints.

![image](https://user-images.githubusercontent.com/11030959/197875999-f20320b1-8db3-4710-8dd2-4ab2825f59af.png)

## Screenshot 3

Here is a screenshot of the configuration for our Model - `Website Suggestions`. It is setup to use the Search Hub named `website-full-search`.

![image](https://user-images.githubusercontent.com/11030959/197875293-8bd67203-ac1e-4bc6-b70c-f08656249499.png)

**Comment by akaustav** [comment]
So far, here is what I have learnt (alongside discussions with some members on the coveo team who liason with our company - Christian Bilodeau (<cbilodeau@coveo.com>) and Adrian Figueroa (<afigueroa@coveo.com>):

1. We are using almost a year old version of `@coveo/headless` (v1.62.0) and `@coveo/angular` (v1.19.0). There is a strong suspicion that we are running into a bug that was corrected with the `@coveo/headless` [v1.85.1](https://github.com/coveo/ui-kit/blob/master/packages/headless/CHANGELOG.md#1851-2022-07-11) where an issue was corrected with the click event payload. This is possibly causing the click analytics events to not fire. This is similar to the call that fires when we click on links in the search results on coveo.com - e.g: `POST https://analytics.cloud.coveo.com/rest/ua/v15/analytics/click?visitor=5a9c6498-56d6-4cc6-8fdc-1614b3205b1c`

2. It is possible that the demo application does NOT already have a sample for triggering the "view" analytics. Similar to what happens on coveo.com when you navigate between pages - on view of each page: `POST https://usageanalytics.coveo.com/rest/v15/analytics/view`

The lack of these two analytics events on our website is possibly preventing the Query Suggest Model from learning anything about user search behaviors.

**Comment by y-lakhdar** [comment]
Thanks for the additional info,
However, I doubt any of these are the root cause to your issue. Specially, if you say that the query suggest "stopped working" after you have moved to Coveo Headless. 

As for the "view" event, this is mainly to feed the [Coveo Machine Learning](https://docs.coveo.com/en/188/glossary/coveo-machine-learning) [event recommendations](https://docs.coveo.com/en/1016/glossary/event-recommendations) [models](https://docs.coveo.com/en/1012/glossary/model).

Here are some questions to help us troubleshoot the issue
1. When you open your query suggest model configuration, are you seeing any learnt suggestions?
2. Does the originel level1 values matches the search hub used in your Headless search page?
3. Are you getting suggestions in the Content Browser (Coveo Admin Console) when you use the pipeline to which the Query Suggest ML Model was associated?

![image](https://user-images.githubusercontent.com/12199712/197793301-f2b5271b-2186-498b-b70e-109dd0449535.png)
![image](https://user-images.githubusercontent.com/12199712/197795651-73fb7abe-10f0-4324-8286-e2d0c679925c.png)

**Reply by akaustav** [reply]
Thanks for your response, @y-lakhdar. My responses are inline. Screenshots below:

> 1. When you open your query suggest model configuration, are you seeing any learnt suggestions?

No learnt suggestions.

> 2. Does the origin level1 values matches the search hub used in your Headless search page?

In our case, the `originLevel1` does NOT even show around the area indicated in your screenshot.

> 3. Are you getting suggestions in the Content Browser (Coveo Admin Console) when you use the pipeline to which the Query Suggest ML Model was associated?

No

## Screenshot about Q1 and Q2

![image](https://user-images.githubusercontent.com/11030959/197882811-4c222721-0238-4bfb-b57f-95db7ab5a199.png)

## Screenshot about Q3

A screenshot showing an empty array for completions and no search suggestions when we click on the search bar within the "Content > Content Browser" section of Coveo Admin Console.

![image](https://user-images.githubusercontent.com/11030959/197882029-a85c6e38-024a-447e-aaeb-813d30154506.png)

**Comment by y-lakhdar** [comment]
Hi @akaustav,

Can you now see the click events being fired in the network tab, when you click on one of your search result (in your Angular app)?
It's unlikely that the headless version you are using would be the reason why click events are not fired.

**Reply by akaustav** [reply]
Hi @y-lakhdar,

> Can you now see the click events being fired in the network tab, when you click on one of your search result (in your Angular app)?

Not at the moment - I don't see the calls to the click events. You can see for yourself by navigating to one of the search result pages on our public facing production website: <https://www.republicservices.com/search/recycling>.

On the other hand, I do see calls to these click events in the sample angular application generated by the latest `@coveo/cli`. Hence, my suspicion on the version of `@coveo/headless` and `@coveo/angular` that we use.

The screenshot below shows the call to the click analytics upon clicking the first link on the search results page i.e. [Lafayette, CO Rates | Republic Services](https://www.republicservices.com/municipality/lafayette-co/rates) (To make the screenshot look cleaner - I opened the link in a different tab - hence, the page is still displaying the search results page).

![image](https://user-images.githubusercontent.com/11030959/198412978-40135e72-09fa-4fcf-bdf3-fa6e70026fe7.png)

Ameet

**Comment by y-lakhdar** [comment, marked as answer]
Hi @akaustav,
This seems to be an implementation issue. And since this is unrelated to the Coveo CLI, I would recommend you to contact Coveo Support.

In the meantime, you can use the Angular app generated by the CLI and compare with your current app.

**Reply by akaustav** [reply]
Thanks @y-lakhdar. I will reach out to coveo support. Meanwhile, the coveo cli related question was that if there is a way to generate an Angular application have examples of other non-standard analytics calls - especially view events.

**Reply by y-lakhdar** [reply]
Not at the moment. Since the current Angular app only focuses on a search page, there is no page view event.
We don't usually want the ML model to recommend a search page.

Here is [some explanation](https://docs.coveo.com/en/1818/leverage-machine-learning/send-usage-analytics-view-events) of how you could log a page view event in your app:
